File: mini-claw-code-starter-py/src/mini_claw_code_starter_py/mock.py
from __future__ import annotations

from collections import deque
from typing import Sequence

from .types import AssistantTurn, Message, Provider, ToolDefinition


class MockProvider(Provider):
    """Chapter 1: implement a canned provider backed by a deque."""

    def __init__(self, responses: deque[AssistantTurn]) -> None:
        self.responses = responses

    @classmethod
    def new(cls, responses: deque[AssistantTurn]) -> "MockProvider":
        raise NotImplementedError("Wrap the response deque in a MockProvider and return it")

    async def chat(
        self,
        _messages: Sequence[Message],
        _tools: Sequence[ToolDefinition],
    ) -> AssistantTurn:
        raise NotImplementedError("Pop the next response from the left or raise if empty")

File: mini-claw-code-starter-ts/src/mock.ts
import type { AssistantTurn, Message, Provider, ToolDefinition } from "./types";

/**
 * Chapter 1 exercise:
 * return canned responses in FIFO order through the Provider interface.
 */
export class MockProvider implements Provider {
  constructor(private readonly responses: AssistantTurn[]) {
    void responses;
  }

  static new(responses: AssistantTurn[]): MockProvider {
    return new MockProvider(responses);
  }

  async chat(
    _messages: Message[],
    _tools: ToolDefinition[],
  ): Promise<AssistantTurn> {
    throw new Error(
      "TODO(ch1): consume the next canned response and throw when none remain",
    );
  }
}
